### PR TITLE
sBt2

### PR DESCRIPTION
**Pull request by LouisaHenze**
In website.py the method input_files was changed to compute_similarities_of_files that computes the cosine similarity of two files a user specified in a web form. The cosine similarity is returned in a textfield of the web form.

Problem: If one of the computed embeddings is None, the cosine similarity cannot be computed. Therefore there is a message in the textfield that the file is probably too large to compute an embedding of it. But if the user did not set the openai api key as environment variable, the method to compute the embedding also returns None. At the moment the string in the textfield just says that the file is probably too large which is not necessarily the case if the key is not set, but you get a message in terminal that the key has to be set as environment variable. Should it be added to the displayed text in the textfield that the openai api key may not be set as environment variable?

### CODE REVIEW

File: src/website.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+from openai.embeddings_utils import cosine_similarity
+
+from compute_embedding import embedding_from_string
+
+app = Flask(__name__)
+
+@app.route('/')
+def home():
+    return '''<form>
+    <label for="fname">Enter first file file: </label><br>
```
Comment (MarioStanke) [inline src/website.py:13]: file file -> file name
This makes sense under limited use cases only. Usually the server and client are different computers.
The server must be able to run Python, with the special libraries and OPEN_API_KEY. The client is just some computer with a browser. Only in this limited development setting the server and client are the same computer. 
The file must be on the server in a particular directory.

File: src/website.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    Returns: A web form is returned that contains two labels file1 and file2 
+        and the names of the files the user specified. In a textfield the 
+        cosine similarity of the content of the two files is displayed. If the
+        embedding for one of the files cannot be computed, there is a message
+        that the file is probably too large in the textfield.     
+    """
+    # Test if files exist.
+    if not ((path.exists(file1) and path.exists(file2))):
+      return "<p>File does not exist</p>"
```
Comment (MarioStanke) [inline src/website.py:69]:
```suggestion
      return "<p>At least one file does not exist on the server.</p>"
```

**Review by MarioStanke — CHANGES_REQUESTED** [review]
If an input file is too large, the app does not respond gracefully. Please cut it at max_tokens instead. For this you have to remove the `return [None]` from `embedding_from_string`.

A user/developer should know what to input, i.e. they have to place a text file where they executed the website.py server starting script. Please add this to the doc string description of the arguments.

There are further issues with the app that Lukasz has to address.

**Comment by LouisaHenze** [comment]
> If an input file is too large, the app does not respond gracefully. Please cut it at max_tokens instead. For this you have to remove the `return [None]` from `embedding_from_string`.
> 
> A user/developer should know what to input, i.e. they have to place a text file where they executed the website.py server starting script. Please add this to the doc string description of the arguments.
> 
> There are further issues with the app that Lukasz has to address.

I cannot only remove `return [None]` from `embedding_from_string`. The problem is that `embedding_from_string` computes the embedding of the string and not of the tokens. Therefore, if the string that should be embedded is too long (because it is encoded in too many tokens) `embedding_from_string` can notice that the string is too long by computing the number of tokens for the string. But you do not get a shorter string that has exactly the maximum number of tokens (the first max number of tokens of tokens of the original string) if it is encoded. Because the string is not changed in `embedding_from_string` the embedding of a string that is too long cannot be computed. 
I would suggest to write a `decode` function in tokenizer.py  and then change  `embedding_from_string` such that it computes with the  `decode` function a shorter string and embedds it.

**Review by MarioStanke — APPROVED** [review]
Ich hatte schon wieder vergessen, dass wir das Embedding direkt nicht von den Tokens bekommen, sondern nur vom String.
Looks good.

**Pull request merged**